File: 0049-group-anagrams/0049-group-anagrams.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        hashmap = collections.defaultdict(list)
        
        for s in strs:
            counts = [0] * 26
            
            for c in s:
                counts[ord(c) - ord('a')] += 1
            
            key = tuple(counts)
            hashmap[key].append(s)
        
        return hashmap.values()

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        // approach: hashmap of char count array : list of strings
        
        Map<String, List<String>> hashmap = new HashMap<>();
        
        for (String str : strs) {
            Integer[] buffer = new Integer[26];
            
            for (int i = 0; i < buffer.length; i++) {
                buffer[i] = 0;
            }
            
            for (int i = 0; i < str.length() ; i++) {
                // update key with char counts
                int index = str.charAt(i) - 'a';
                buffer[index]++;
            } 
            
            String key = new String(Arrays.toString(buffer));
            List<String> list = hashmap.computeIfAbsent(key, k -> new ArrayList<>());
            list.add(str);
        }
        
        return new ArrayList<>(hashmap.values());
    }
}